eZ/Publish/Core/Search/Legacy/Content/Handler.php

```diff
@@ -193,6 +193,7 @@ public function findContent(Query $query, array $languageFilter = array())
 
     protected function extractMatchedLanguage($languageMask, $mainLanguageId, $languageSettings)
     {
+        // @todo load all at once or better yet load several at once? Either way check if this is using cached instance
```

Comment:
I'd say bulk load all Ids that are in `$languageSettings['languages']`.
Side note: there's `\eZ\Publish\Core\Persistence\Legacy\Content\Language\MaskGenerator` injectable as `@ezpublish.persistence.legacy.language.mask_generator`. Maybe entire responsibility should be moved there?

Reply:
Maybe, I'll move this to separate PR